### PR TITLE
Optional field nullable issue fix

### PR DESCRIPTION
**Pull request by nitesh-idnani1**
Fixed null value update in mongo db by changing dotnotation logic for null object and added test for the same

### CODE REVIEW

File: papiea-client/__tests__/update.test.ts [hunk 1]
```diff
@@ -6,11 +6,11 @@ import { IntentfulBehaviour, IntentfulStatus } from "papiea-core"
 declare var process: {
     env: {
         SERVER_PORT: string,
-        PAPIEA_ADMIN_S2S_KEY: string
+        PAPIEA_ADMIN_KEY: string
```
Comment (shlomi-nx) [inline papiea-client/__tests__/update.test.ts:9]: Why did you rename this? What was wrong with `PAPIEA_ADMIN_S2S_KEY`?

**Review by shlomi-nx — COMMENTED** [review]
Why did we rename `PAPIEA_ADMIN_S2S_KEY`?

Unless you convince me otherwise, I'd like to keep using the original name

File: papiea-client/__tests__/update.test.ts (hunk 1 above)
Comment (nitesh-idnani1) [reply papiea-client/__tests__/update.test.ts:9]: Since the name was already changed in [docker-compose](https://github.com/nutanix/papiea/blob/2651307fedc4ce420d589a92d99a96e7bd8b76ef/papiea-engine/docker-compose.yml#L14) file, we decided to change it in the tests as well and keep things consistent.

Comment (shlomi-nx) [reply papiea-client/__tests__/update.test.ts:9]: It shouldn't have been changed there. Please revert this change.

Comment (nitesh-idnani1) [reply papiea-client/__tests__/update.test.ts:9]: We should be changing the name in `docker-compose.yml` file as well right i.e. keep PAPIEA_ADMIN_S2S_KEY everywhere?

Comment (shlomi-nx) [reply papiea-client/__tests__/update.test.ts:9]: Yes, thanks!

File: papiea-engine/__tests__/provider_tests/provider_api.test.ts [hunk 2]
```diff
@@ -269,19 +268,14 @@ describe("Provider API tests", () => {
...
-            expect(e.response.data.error.message).toEqual("Error parsing update query. Update body might be 'undefined', if this is expected, please use 'null'.")
-        }
+        await providerApi.patch(`/${ provider.prefix }/${ provider.version }/update_status`, {
+            context: "some context",
+            entity_ref: {
+                uuid: metadata.uuid,
+                kind: kind_name
+            },
+            status: newStatus
+        });
```
Comment (shlomi-nx) [inline papiea-engine/__tests__/provider_tests/provider_api.test.ts:278]: Where are the tests and assertions here? This is an empty test.

File: papiea-engine/docker-compose.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ services:
                 ports:
                         - "3333:3000"
                 environment:
-                        - PAPIEA_ADMIN_KEY=${PAPIEA_ADMIN_S2S_KEY}
+                        - PAPIEA_ADMIN_S2S_KEY=${PAPIEA_ADMIN_S2S_KEY}
```
Comment (shlomi-nx) [inline papiea-engine/docker-compose.yml:14]: Please verify that this change is reflected in any .env file (if there are any) as well as in any env settings such as Circle CI etc.

File: papiea-engine/src/provider/provider_api_impl.ts [hunk 4]
```diff
@@ -7,12 +7,13 @@ import { Authorizer } from "../auth/authz";
...
 import { createHash } from "../auth/crypto";
 import { EventEmitter } from "events";
-import { Action, Entity_Reference, Provider, S2S_Key, Secret, Status, Version } from "papiea-core";
+import { Action, Entity_Reference, Kind, Provider, S2S_Key, Secret, Status, Version } from "papiea-core";
 import { Logger } from "papiea-backend-utils";
 import { Watchlist_DB } from "../databases/watchlist_db_interface";
 import { SpecOnlyUpdateStrategy } from "../intentful_core/intentful_strategies/status_update_strategy";
 import { IntentfulContext } from "../intentful_core/intentful_context";
 import uuid = require("uuid");
+import { isEmpty } from "lodash";
```
Comment (shlomi-nx) [inline papiea-engine/src/provider/provider_api_impl.ts:16]: Where is `isEmpty` being used?

File: papiea-engine/src/provider/provider_api_impl.ts [hunk 5]
```diff
@@ -132,6 +136,18 @@ export class Provider_API_Impl implements Provider_API {
         return await this.providerDb.get_latest_provider_by_kind(kind_name);
     }
 
+    async fix_null_object(schemas: any, status: Status): Promise<Status> {
```
Comment (shlomi-nx) [inline papiea-engine/src/provider/provider_api_impl.ts:139]: Why is this function async?

File: papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts [hunk 6]
```diff
@@ -926,6 +927,53 @@ describe("Provider Sdk tests", () => {
             sdk.server.close();
         }
     });
+
+    const null_test_yaml = load(readFileSync(resolve(__dirname, "../test_data/null_test_data.yml"), "utf-8"));
+    test("Test entity initialization after setting it to null", async () => {
+        expect.hasAssertions();
```
Comment (shlomi-nx) [inline papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts:933]: Its probably good to list the amount of assertions in expected in this function.

File: papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts [hunk 7]
```diff
@@ -926,6 +927,53 @@ describe("Provider Sdk tests", () => {
...
+                        tenant_id: '1',
+                        account_id: '2'
+                    }
+                }
+            });
+
+            const res: any = await axios.post(`${sdk.entity_url}/${sdk.provider.prefix}/${sdk.provider.version}/${kind_name}/${metadata.uuid}/procedure/testProcedure`, { input: '2' });
+            const updatedEntity: any = await axios.get(`${sdk.entity_url}/${sdk.provider.prefix}/${sdk.provider.version}/${kind_name}/${metadata.uuid}`);
+            expect(updatedEntity.data.status.provider_ref.tenant_id).toEqual('1')
+            expect(updatedEntity.data.status.provider_ref.account_id).toEqual('2')
```
Comment (shlomi-nx) [inline papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts:970]: e.g. null dereference in the second `expect` would make this test pass.

File: papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts [hunk 8]
```diff
@@ -926,6 +927,53 @@ describe("Provider Sdk tests", () => {
...
+        } catch(e) {
+            console.log(e.response.data.error)
```
Comment (shlomi-nx) [inline papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts:972]: unless you fail the test explicitly here, which may not be a bad idea.

**Review by shlomi-nx — COMMENTED** [review]
Please see my comments below

File: papiea-engine/__tests__/provider_tests/provider_api.test.ts (hunk 2 above)
Comment (nitesh-idnani1) [reply papiea-engine/__tests__/provider_tests/provider_api.test.ts:278]: Adding expect statements

File: papiea-engine/docker-compose.yml (hunk 3 above)
Comment (nitesh-idnani1) [reply papiea-engine/docker-compose.yml:14]: Ran this on CircleCI, everything ran successfully

File: papiea-engine/src/provider/provider_api_impl.ts (hunk 4 above)
Comment (nitesh-idnani1) [reply papiea-engine/src/provider/provider_api_impl.ts:16]: Fixed

File: papiea-engine/src/provider/provider_api_impl.ts (hunk 5 above)
Comment (nitesh-idnani1) [reply papiea-engine/src/provider/provider_api_impl.ts:139]: Fixed this

File: papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts (hunk 6 above)
Comment (nitesh-idnani1) [reply papiea-sdk/typescript/__tests__/sdk_tests/provider_sdk.test.ts:933]: Fixed this

**Review by shlomi-nx — APPROVED** [review]
